ESPeasy codemirror

chromoxdor opened the pull request:
I am struggling with the links of the files since i get an error message. For now they have no path, so the must be available on the eps itself.
I am also pretty sure it is not the right place for those links.  

Fixes: https://github.com/letscontrolit/ESPEasy/issues/4086

chromoxdor committed:
initial codemirror support

chromoxdor committed:
back to local testing

chromoxdor commented on src/src/WebServer/WebServer.cpp:
```diff
@@ -408,6 +408,10 @@ void getWebPageTemplateDefaultHead(WebTemplateParser& parser, bool addMeta, bool
   if (parser.isTail()) return;
   parser.process(F("<!DOCTYPE html><html lang='en'>"
             "<head>"
+            "<link rel='stylesheet' href='codemirror.css'>"
```
i guess this all must be in "WebStaticData" but this is beyond my capabilities... 😳

tonhuisman replied:
It should have a `/` before that last `>` to make that `link` tag valid.

And the `script` tags below can be simplified by replacing `></script>` by `/>`

chromoxdor replied:
Thanks... 

> It should have a `/` before that last `>` to make that `link` tag valid.

This works, there is no / missing (https://www.w3schools.com/css/css_howto.asp)

tonhuisman replied:
Ah, then that's one of the peculiarities (read: diversions) from the regular XML-style rules that apply to the HTML dialect 💇‍♂️

tonhuisman replied:
> And the script tags below can be simplified by replacing `></script>` by `/>`

You can scrape off a few bytes of the .bin size by applying this change (3x).

Would it work if you added an `#ifdef` around this part for conditional inclusion, to be able leave it out of the 'minimal' builds?

chromoxdor replied:
Codemirror is in my opinion to big to be ever integrated in an espeasy-binary. 
So i think the links to the files should be always there regardless of the kind of build. And the acutal files can be either served  online or locally (if present).  It actually also doesn’t need a check if the files are available at all, since if not, it simply falls back to the normal textinput 

That´s what @TD-er wanted to do because i am not able to do so (or at least it will take me a long time to become familiar with c++ 😳 ) 
And the webserver.cpp isn’t the right place anyway... i guess

chromoxdor committed:
minor change

chromoxdor committed:
Update espeasy.js

TD-er commented on src/src/WebServer/Rules.cpp:
```diff
@@ -552,6 +552,7 @@ void Rule_showRuleTextArea(const String& fileName) {
   addHtml(F("<textarea id='rules' name='rules' rows='30' wrap='off'>"));
   size = streamFromFS(fileName, true);
   addHtml(F("</textarea>"));
+  addHtml(F("<script>initCM();</script>"));
```
There is a function to add scripts and also to serve it directly where it checks to see if the static file is included during build and if not, it will serve it via the CDN links
I will try to add it for you, pushing on top of this PR.

chromoxdor replied:
Great...

TD-er commented:
I am missing the minified .js and .css files.

chromoxdor commented:
oh..i see i only minified the codemirror.js
on it

chromoxdor committed:
added minified files

TD-er commented:
Later we could also add some extra GET URL which will generate the available commands for hinting based on what is included in the build.
This also allows to color some valid keywords which may not work on the current build.

But that's for later.

chromoxdor commented:
hmmm.. not a bad idea. but even without individualized wordlist, this a huge improvement in writing rules. especially for people like me that always forget the commands. :)
i copied all the commands from the notepad++ repo of grovkillen and added all math & number related stuff.
If something is missing let me know....

tonhuisman commented:
> If something is missing let me know....

I'm not sure if all plugin commands are included in there, especially for the more recently added plugins?

Might need a separate line of these commands per plugin to make merging easier, as git doesn't play nice when only a part of a line is changed, and _does_ seem to handle separate lines just fine.

chromoxdor commented:
> I'm not sure if all plugin commands are included in there, especially for the more recently added plugins?

i will look at all the plugins the next days/weeks and add missing things...

chromoxdor committed:
update

-centralized "wordlist"
> keywords now need only to be in espeasy.js

-caseinsenitive search for known keywords

tonhuisman commented on static/espeasy.js:
```diff
@@ -1,11 +1,41 @@
+
+var commonAtoms = ["And", "Or"];
+var commonKeywords = ["If", "Else", "Elseif", "Endif"];
+var commonCommands = ["AcessInfo", "Background", "Build", "ClearAccessBlock", "ClearRTCam", "Config", "ControllerDisable",
```
Looks like "AcessInfo" has a typo, according to the code it should be "AccessInfo"

tonhuisman commented on static/espeasy.js:
```diff
@@ -1,11 +1,41 @@
+  "TaskDisable", "TaskEnable", "TaskRun", "TaskValueSet", "TaskValueSetAndRun", "TimerPause", "TimerResume", "TimerSet", "TimerSet_ms", "TimeZone",
+  "UdpPort", "UdpTest", "Unit", "UseNTP", "WdConfig", "WdRead", "WifiAPkey", "WifiAllowAP", "WifiAPMode", "WifiConnect", "WifiDisconnect", "WifiKey",
+  "WifiKey", "WifiScan", "WifiSSID", "WifiSSID2", "WifiSTAMode",
+  "Event", "AsyncEvent", "/control?cmd",
```
I'm not sure if "/control?cmd" should be a suggestion or keyword, IMHO it shouldn't be in here.

chromoxdor replied:
Thanks.. i copied these words blindly from the notepad++ version.. i also asked myself why "/control?cmd"  is in the list 🤷🏻‍♂️

tonhuisman replied:
And the second "WifiKey" instance should probably read "WiFiKey2" (spelling the word WiFi must be difficult, for whoever created the original list 😛)

chromoxdor replied:
> > I'm not sure if "/control?cmd" should be a suggestion or keyword, IMHO it shouldn't be in here.

i am going through all the plugins right now...
thats why: https://espeasy.readthedocs.io/en/latest/Plugin/P091.html#p091-page

tonhuisman replied:
Well, that's just *one* way of activating that command, not **the** way, nor the generic way 😉

tonhuisman commented on static/espeasy.js:
```diff
@@ -1,11 +1,41 @@
+  "MonitorRange,PCF", "UnMonitorRange,PCF", "PCFGPIORange", "PCFMode", "Tone", "RTTTL", "UnMonitor",];
+var commonString2 = ["Clock#Time", "Login#Failed", "MQTT#Connected", "MQTT#Disconnected", "MQTTimport#Connected", "MQTTimport#Disconnected", "Rules#Timer", "System#Boot",
+  "System#BootMode", "System#Sleep", "System#Wake", "TaskExit#", "TaskInit#", "Time#Initialized", "Time#Set", "WiFi#APmodeDisabled", "WiFi#APmodeEnabled",
+  "WiFi#ChangedAccesspoint", "WiFi#ChangedWiFichannel", "WiFi#Connected"];
```
There is also a "WiFi#Disconnected" event, not (yet) listed here.

tonhuisman commented on static/espeasy.js:
```diff
@@ -1,11 +1,41 @@
+var commonTag = ["On", "Do", "Endon"];
+var commonNumber = ["toBin", "toHex", "Constrain", "XOR", "AND:", "OR:", "Ord", "bitRead", "bitSet", "bitClear", "bitWrite", "urlencode"];
+var commonMath = ["Log", "Ln", "Abs", "Exp", "Sqrt", "Sq", "Round", "Sin", "Cos", "Tan", "aSin", "aCos", "aTan", "Sind_d", "Cos_d", "Tan_d", "aSin_d", "aCos_d", "sTan_d"];
+var commonWarning = ["delay", "Delay", "Yehudit"];
```
Feels like "Yehudit" is some personal setting or preference (copied over from the other highlighter?)

chromoxdor replied:
ubbs 😬

tonhuisman commented on static/espeasy.js:
```diff
@@ -1,11 +1,41 @@
+var commonMath = ["Log", "Ln", "Abs", "Exp", "Sqrt", "Sq", "Round", "Sin", "Cos", "Tan", "aSin", "aCos", "aTan", "Sind_d", "Cos_d", "Tan_d", "aSin_d", "aCos_d", "sTan_d"];
+var commonWarning = ["delay", "Delay", "Yehudit"];
+var AnythingElse = [];
+
```
As a general remark on these lists: It will be helpful to group the keywords, f.e. by topic or plugin, and put them on a separate line (and ending with a comma already, ts/js/json allows that), with the closing `];` also on a separate line, to make merging in extra/new (plugin-specific) keywords easier. Then it is more attractive for developers to add new keywords to the lists, as they will then also be highlighted in the editor. 😉

chromoxdor replied:
Good, idea... therefore i need to figure out which keyword of the existing list belongs to which plugin

but you mean it like this?:

```
var commonString2 = [
//plugin123
"Clock#Time", "Login#Failed", "MQTT#Connected", "MQTT#Disconnected", "MQTTimport#Connected", 
//plugin234
"MQTTimport#Disconnected", "Rules#Timer", "System#Boot",
 ];
```

chromoxdor committed:
fixed typo & removed an unintentional easteregg

chromoxdor committed:
each plugin one line

chromoxdor committed:
Merge branch 'letscontrolit:mega' into esp_codemirror

chromoxdor commented:
Just tried the action build....I realized, that my other devices where discovered immediately. Is the P2P Plugin now activated by default?

tonhuisman commented:
> Is the P2P Plugin now activated by default?

No, but the Main tab shows any device reporting the correct ESPEasy info via the UDP port configured. It can also list RPIEasy devices there 😉 Just be sure to have unique Unit IDs configured.

chromoxdor committed:
Forgot to rename variable in anyword

chromoxdor commented:
Ahh i see... and that works because now a default udp port is set...

tonhuisman commented:
> and that works because now a default udp port is set...

That's probably why I haven't used that default port, and I thought I might want to create separate independent groups of ESPs, but I still haven't 😅

chromoxdor committed:
typo: commonPlugins

TD-er commented:
Just curious.
The commands and events in the JavaScript files, are they case-insensitive when parsed for color code highlighting?
Events and commands are processed case insensitive.
But for readability, it is better to have them being suggested for tab completion with the case they are now present in the files of this PR.

chromoxdor commented:
The wordlist is merged with an lower case version of itself... so for code highlighting you can write as you like. 
the hints appear as we write them into the array.. you can start typing case insensitive but the autocorrection will turn them into case sensitiv words...

Edit: i just basically wrote the same sentences as you :) so for short: Yes it is like you said.
Edit2: The default of codemirror was case sensitive for everything. It took me a while since i am so bad at JS but i did a lot of trial&error because stackoverflow (writing and waiting for answers) is to slow for me :)

chromoxdor committed:
added missing plugin commands + code cleanup

chromoxdor commented:
@tonhuisman  can you help me with P116. I am not sure with the commands...

chromoxdor commented:
if it would be possible to retrieve a list of all the plugins as numbers installed on a device i could integrate a plugin dependent hinting. e.g. added in the regular json output

tonhuisman commented:
Clippings from the documentation:
//P123 (FT6206 touch screen:
```
touch,rot,<0..3>
touch,flip,<0|1>
```
//ESPEasy_TouchHandler: (Used for P123, and later also for P099)
```
touch,enable,<objectName|Nr>[,...]
touch,disable,<objectName|Nr>[,...]
touch,on,<buttonObjectName|Nr>[,...]
touch,off,<buttonObjectName|Nr>[,...]
touch,toggle,<buttonObjectName|Nr>[,...]
touch,setgrp,<group>
touch,incgrp
touch,decgrp
touch,incpage
touch,decpage
touch,updatebutton,<buttonName|Nr>[,<group>[,<mode>]]
```
//P095 ILI9xxx, P096 eInk, P116 ST77xx, P131 NeoPixelMatrix:
`<trigger>` is the command 'prefix'.
For P095 the user can select one of: `tft`, `ili9341`, `ili9342`, `ili9481`, `ili9486`, `ili9488`.
For P096 the user can select one of: `eink`, `epaper`, `il3897`, `uc8151d`, `ssd1680`, `ws2in7`, `ws1in54`.
For P116 the user can select one of: `tft`, `st77xx`, `st7735`, `st7789`, `st7796`.
For P131 the user can select one of `neo`, `neomatrix`.
For plugin-specific commands the suffix `cmd` is added.
```
<trigger>cmd,on // f.e. tftcmd,on
<trigger>cmd,off // On and Off not for P131
<trigger>cmd,clear
<trigger>cmd,backlight,<percentage> // not for P096 and P131
<trigger>cmd,bright,<percentage> // Only for P131
```
// P096 only:
```
<trigger>cmd,deepsleep
<trigger>cmd,seq_start
<trigger>cmd,seq_end
```
// P095 and P096:
```
<trigger>cmd,inv  // Backward compatibility
<trigger>cmd,rot,<rotation> // Backward compatibility
```
// AdafruitGFX_Helper commands: (Used for P095 (ILI9xxx), P096 (eInk), P116 (st77xx), P131 (NeoPixelMatrix)
```
<trigger>,clear[,<color>]
<trigger>,rot,<rotation>
<trigger>,tpm,<TextPrintMode>
<trigger>,txt,<text>
<trigger>,txp,<x>,<y>
<trigger>,txz,<x>,<y>,<text>
<trigger>,txc,<foreColor>[,<backgroundColor>]
<trigger>,txs,<size>
<trigger>,txtfull,<x>,<y>,<size=1>,<foreColor=white>,<backColor=transparent>,<text>
<trigger>,asciitable[,<offset>,<fontscaling>]
<trigger>,font,<fontname>
<trigger>,l,<x1>,<y1>,<x2>,<y2>,<color>
<trigger>,lh,<y>,<width>,<color>
<trigger>,lv,<x>,<height>,<color>
<trigger>,lm,<color>,<x1>,<y1>,<x2>,<y2>[,<colorN>][,<xN>,<yN>[,...]|c]
<trigger>,lmr,<color>,<x>,<y>,<dx1>,<dy1>[,<colorN>][,<dxN>,<dyN>[,...]|c]
<trigger>,r,<x>,<y>,<width>,<height>,<color>
<trigger>,rf,<x>,<y>,<width>,<height>,<bordercolor>,<innercolor>
<trigger>,c,<x>,<y>,<radius>,<color>
<trigger>,cf,<x>,<y>,<radius>,<bordercolor>,<innercolor>
<trigger>,t,<x1>,<y1>,<x2>,<y2>,<x3>,<y3>,<color>
<trigger>,tf,<x1>,<y1>,<x2>,<y2>,<x3>,<y3>,<bordercolor>,<innercolor>
<trigger>,rr,<x>,<y>,<width>,<height>,<corner_radius>,<color>
<trigger>,rrf,<x>,<y>,<width>,<height>,<corner_radius>,<bordercolor>,<innercolor>
<trigger>,px,<x>,<y>,<color>
<trigger>,pxh,<x>,<y>,<color1>[,<color2>,...]
<trigger>,pxv,<x>,<y>,<color1>[,<color2>,...]
<trigger>,bmp,<x>,<y>,<filename.bmp>
<trigger>,btn,<state>,<mode>,<x>,<y>,<w>,<h>,<id>,<type>,<ONcolor=blue>,<OFFcolor=red>,<CaptionColor=white>,<fontsize>,<ONcaption>,<OFFcaption>,<BorderColor=white>,<DisabledColor=0x9410>,<DisabledCaptionColor=0x5A69>,<TaskIndex>,<Group>,<SelectGroup>,<objectname>
```

Sorry for the long list... 🧁

chromoxdor commented:
>\<trigger>cmd,on // f.e. tftcmd,on
> \<trigger>cmd,off // On and Off not for P131
> \<trigger>cmd,clear
> \<trigger>cmd,backlight,<percentage> // not for P096 and P131
> \<trigger>cmd,bright,<percentage> // Only for P131

Is there a seperator missing between \<trigger> & cmd?



> Sorry for the long list... 🧁

😳
Maybe i have to find a new way to handle all this commands in terms of the current keyword structure, since there are many that are used by different plugins....

tonhuisman commented:
> > `<trigger>cmd,on // f.e. tftcmd,on`
> 
> Is there a seperator missing between <trigger> & cmd?

No, that's why I included an example, I expected that to be confusing, but it is backward compatible with how it was initially created 😃 
So there are commands that start with `tft` and also a few commands that start with `tftcmd`. And if another trigger is selected in the device settings, the commands could be `st77xx` and `st77xxcmd`.

NB: Highlighting only for available plugins isn't really needed, I'd think, that would make the highlighter code really messy and complicated.
NB2: The overlap in commands is to reduce code for similar plugins, like the displays (I have a few more different hardware types waiting to be implemented...) and touch screens (that will probably stop after P099 and P123, as I haven't found other touch screen devices yet)

chromoxdor commented:
> NB: Highlighting only for available plugins isn't really needed, I'd think, that would make the highlighter code really messy and complicated.

No it is about hints. And that could be easily integrated... but with or without it is anyway a big change compared to the simple texteditor. Especially when you have to write rules for your latest plugins 😁

chromoxdor committed:
added all display specific commands

chromoxdor commented:
@tonhuisman i added all the display commands. therefore i added two new dispay specific arrays to iterate to each displaytype and add the subcommands....

chromoxdor committed:
Merge branch 'mega' into esp_codemirror

chromoxdor committed:
added , for hinting

since there are so many display commands that contain a "," i added it to the hinting for easier use...

chromoxdor committed:
Merge branch 'esp_codemirror' of https://github.com/chromoxdor/ESPEasy into esp_codemirror

chromoxdor committed:
removing duplicates in hints

chromoxdor committed:
Merge branch 'mega' into esp_codemirror

chromoxdor committed:
nobody except me needs the log

chromoxdor committed:
Merge branch 'esp_codemirror' of https://github.com/chromoxdor/ESPEasy into esp_codemirror

chromoxdor committed:
Merge branch 'mega' into esp_codemirror

chromoxdor committed:
Merge branch 'letscontrolit:mega' into esp_codemirror

chromoxdor committed:
bit more "root" keywords for faster typing

chromoxdor committed:
maybe better hint listing

chromoxdor committed:
Update anyword-hint.min.js

chromoxdor committed:
changes

chromoxdor committed:
sevenseg fix

chromoxdor committed:
Merge branch 'letscontrolit:mega' into esp_codemirror

chromoxdor committed:
added files

chromoxdor commented:
I updated this branch.
A folder with all the necessary files was added.
You still have to copy the files to the esp....

tonhuisman commented:
> You still have to copy the files to the esp....

Great!
Can you document that process in a `readme.md`, or similar file, in the same folder? As that won't be very intuitive to most users.

chromoxdor commented:
i can do that... but the goal would be to direct to a cdn and make local files only optional..like the css in 1m builds.
I you do not want to bother with copying files you can use  my other branch where the files are fetched from jsdelivr...
https://github.com/chromoxdor/ESPEasy/tree/powercolor.

i maybe also better start to learn c++ to integrate it properly

TD-er commented:
> i maybe also better start to learn c++ to integrate it properly

Yep, that's the true reason I did not fix it for you.
Absolutely nothing to do with me forgetting, for sure ;)

TD-er committed:
Merge branch 'mega' into esp_codemirror

TD-er committed:
Merge branch 'mega' into esp_codemirror

TD-er committed:
[Build_CDN] Generate CDN prefix URL based on Git tag

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into esp_codemirror

TD-er committed:
Merge branch 'mega' into esp_codemirror

TD-er commented:
There is a bug in the codemirror stuff when editing rules on a mobile.
If you build using the latest code of this PR, you'll see that it is impossible to type in rules code via mobile.

Where the desktop browser nicely does complete and formats something like this:
![image](https://user-images.githubusercontent.com/3751318/187414372-ff0ab8b6-38bd-48b9-b7cb-0118a0f2cf70.png)

On a mobile you can see the cursor jump a bit, but then ends up right after the "do", when pressing enter.

Edit:
When switching the "desktop" view on the mobile it is possible to edit as code completion suggestions are actually shown.
However the view is so far zoomed in that it's hardly usable.
Also the "Endon" is not appended, even though I did click the suggested "Do".
It immediately shows an enormous list of suggestions for completion.
Not only enormous regarding the nr of items, but also the zoomed font size.

N.B. I did update this branch, so you may need to perform a `git fetch` first and then `git pull` on this branch.

TD-er commented:
Just tested and it also occurs when typing on your [test page](https://raw.githack.com/chromoxdor/EasyColorCode/main/colorcode.html) on mobile

chromoxdor commented:
Interesting… this doesn’t happen on my phone. i’ll look at it as soon as the internet is back. our provider has some issues and it seem like it will take the whole day 😳

chromoxdor commented:
which browser do you use on your mobile?

chromoxdor commented:
just found this: 

> I suspect this is because the autocompletion is hooked up to key bindings, and Android doesn't fire useful key events, making key bindings mostly non-functional.

https://github.com/codemirror/CodeMirror/issues/5182

I guess i need to find a workaround.

TD-er commented:
> which browser do you use on your mobile?

Chrome and according to [howbigismybrowser.com](http://howbigismybrowser.com/) it shows as 432x834

No completion:
![Mobile_no_completion](https://user-images.githubusercontent.com/3751318/187426922-e499833f-9c6c-4d9b-b940-532f07a542b2.gif)

Zoomed in way too far to be useable:
![Mobile_too_big](https://user-images.githubusercontent.com/3751318/187427399-a7c29058-6e6d-48a9-9ef0-a5335b78fe77.gif)

TD-er commented:
> I guess i need to find a workaround.

Or maybe disable code completion on Android/mobile?
Just code highlighting?

Or what do I need to do to not enable code completion? Then we can make it a setting.

chromoxdor commented:
i did a bit more digging (using slow internet with a mobile really really sucks!) it seems more a chrome related issue. i used my android tablet with firefox and it worked fine. with chrome i could reproduce the issue as you described. 
could you check if it is the same for you?
eta for internet is 18:00 .. i will answer the questions in the other pr then…

chromoxdor committed:
update CM files

TD-er commented:
Ah you got reconnected to da intewebz

What did you change? Just an update of CodeMirror?

chromoxdor commented:
yes. i am back :) 

> Just an update of CodeMirror?

yes...

chromoxdor committed:
Merge branch 'letscontrolit:mega' into esp_codemirror

chromoxdor committed:
Merge branch 'letscontrolit:mega' into esp_codemirror

TD-er committed:
[Build] Add FEATURE_RULES_EASY_COLOR_CODE build option

This way it is easier to reduce build size by not including the Easy Color Code feature.

TD-er commented:
Just added a commit to hopefully make it fit again.
It now has a `FEATURE_RULES_EASY_COLOR_CODE` define to toggle whether it is included.
Also it does make CDN URLs based on the static files in the ESPEasy repository.

chromoxdor commented:
Great.. but i now get confused because i have two pending PRs regarding colorcode.... maybe at one point i should close one. So you wanna use this one?

TD-er commented:
If you like, I can merge this one and then we can work our way through the other one to see what else is missing.

You did merge the `mega` branch into this one, so I thought this was the "active" one.

chromoxdor commented:
> You did merge the `mega` branch into this one, so I thought this was the "active" one.

I see... This was actually an accident. 😁 


> If you like, I can merge this one and then we can work our way through the other one to see what else is missing.

I think nothing is missing here but in this PR the codemirror related files are not up to date...
i´ll do this now.  And then we can stay with this PR. The only issue i found is, that the codemirror.min.css is not loaded from FS if present...

chromoxdor committed:
update codemirror files

TD-er committed:
Merge branch 'mega' into esp_codemirror

chromoxdor committed:
Update codemirror.min.js

used an external minifier...
old file: 252kb
new: 173kb

now the esp is way more responsive  when using this file locally

chromoxdor commented:
i think there should be also a disable option for where the colorcode might cause trouble....

TD-er committed:
[CodeMirror] Fix loading code mirror CSS from filesystem

TD-er committed:
Merge remote-tracking branch 'chromoxdor/esp_codemirror' into esp_codemirror

chromoxdor committed:
better minification

chromoxdor committed:
Merge branch 'esp_codemirror' of https://github.com/chromoxdor/ESPEasy into esp_codemirror

TD-er commented:
I added a patch to also allow loading the code mirror CSS from the file system.

I'm now looking into a way to not load the code mirror stuff when not needed, since those files can be quite large and serving them always on any page load from the file system can be quite resource consuming.
So adding a setting to disable loading it may be a good idea too.

chromoxdor commented:
> I added a patch to also allow loading the code mirror CSS from the file system.

Perfect... just tested it. Works fine. 

> I'm now looking into a way to not load the code mirror stuff when not needed,

I agree... as an external source it is almost not noticeable but with loading the files from the local FS it slows things down a lot

TD-er commented:
This makes me think, shouldn't we serve the CSS and JS files separately with some set TTL instead of wrapping it in the same HTML code?
This way the browser may decide whether we need the file.
Or is this a bad idea since browsers may not like to load external files from non-HTTPS sources these days?

TD-er commented:
I was thinking about the following:
Where I now call `streamFromFS` I could generate a link to the file with some prefix (e.g. file size since we don't have a timestamp on the filesystem)
Then when serving that file with the prefix, we strip out the prefix and serve the file. However serving those static files (thus the one with the specific prefix) we add like a `cache-control` header indicating a `max_age`.

chromoxdor commented:
I had just a check in mind..like: if the page we visit is the rules page and we turned colorcode on then put the files in the header.
And since 99% of the user will most likely not store the files locally the lag in calling the rulespage once is negligible.

Edit: The problem i have is, that i can not imagine what each solution means codewise (e.g. size and effort) so it is hard for me to help with decisions...

TD-er commented:
Already working on it...
Fetching the file size to make sure we're not missing a file update is a bit time consuming.
Computing the hash is also time consuming.
Therefore I would like to either add the timestamp when the file presence cache was reloaded, or some random value which remains the same until we clear the cache again.

The problem with using a timestamp is that it might not be as random as you would expect as booting may take roughly the same amount of time, thus basing it on millis() isn't such a good idea.
Also I don't know what to pick when the system time is not (yet) set, otherwise the unix time would be great.
So I guess a random value when the time is not set, or else the unix time if set.

TD-er committed:
[CDN] Serve static files with cache-control from file system

TD-er commented:
Hmm that took quite a while to get it somewhat working.
The ESP will now reply with a 304 (not modified) if a file is still the same as last time.
However, the browser (both Firefox and Chrome) do make the calls and don't serve it directly from its own cache.
No idea why not.

But I would like to know if it is working faster now when serving those large files from the local file system.

chromoxdor commented:
> we strip out the prefix and serve the file

the "stripping out part" doesn´t work
![Bildschirmfoto 2022-09-21 um 19 49 45](https://user-images.githubusercontent.com/33860956/191575586-8aaf1c25-ede1-43c9-b27f-fa987e3e0729.png)

TD-er commented:
It should strip it out when you try to access the file itself.
It must be in the link, but when serving the file from the file system, the prefix is stripped from it to get the actual file from the file system.

See here:
The file was initially fetched (thus a 200 HTTP code) and then on clicking on the next tab it got a 304 (not modified) HTTP code:
![image](https://user-images.githubusercontent.com/3751318/191577210-12de237b-7c66-4492-a48b-0639323e32e2.png)

chromoxdor commented:
Here the files are prefixed from the beginning. So codemirror doesn’t get initialized.  I have all the files on the local FS. 
So when i go to the rules tab the files shouldn’t be prefixed initially, right...

chromoxdor commented:
Wait, that was not the question... 

> But I would like to know if it is working faster now when serving those large files from the local file system.

Yes! 🙂

TD-er commented:
> Here the files are prefixed from the beginning. So codemirror doesn’t get initialized. I have all the files on the local FS. So when i go to the rules tab the files shouldn’t be prefixed initially, right...

Nope, the prefix is only when linking the files.

The idea is that by prefixing it, **only in the URL**, the browser may think it is a different file when the prefix has changed and thus not try to load it from the cache.
The files on the file system should be like they always have been.

tonhuisman commented:
Maybe the, quite minimal, changes from #4233 can also be handled in this PR?

TD-er committed:
[build] Disable normal_alt_wifi_ESP8266_1M due to bin size

TD-er commented:
Are you done with the .js files for now?
When I merge them, it might be a bit "funky" cache related as they will be served from the CDN.

chromoxdor commented:
> Maybe the, quite minimal, changes from #4233 can also be handled in this PR?

I can put them in here but why not just merging #4233?


> Are you done with the .js files for now?

Thats a question i´d like to pass on to @tonhuisman. Are there recent developments/changes of plugin commands? 
I will add P023 today..(just found out that the commands are only in the old documentation)

I also just realized that since the introduction of variables in the espeasy css and the ability to choose a mode i also need to integrate this into the codemirror.css too 🙄

Or we leave it out completely and give the ruleseditor always the same color which would look like this:
<img width="394" alt="Bildschirmfoto 2022-09-22 um 10 03 47" src="https://user-images.githubusercontent.com/33860956/191692321-6d4394d1-0adb-4fc8-b532-1b4e78addf96.png">

and here with following the colorscheme:
<img width="394" alt="Bildschirmfoto 2022-09-22 um 10 03 20" src="https://user-images.githubusercontent.com/33860956/191692504-48ac6f90-fda9-438b-ae96-eb55ebdcf6b4.png">

TD-er commented:
> I can put them in here but why not just merging #4233?

Not 100% sure how that will work out with merging this one later.
Anyway, whether it has to be done manual or via merging, it has to be done... Will do it after posting this :)

> I also just realized that since the introduction of variables in the espeasy css and the ability to choose a mode i also need to integrate this into the codemirror.css too 🙄

I think it is better to have it more consistent, so please have a look to what is needed here.
Then we will later this evening decide whether to merge this PR now or wait a bit more.

TD-er committed:
Merge branch 'mega' into esp_codemirror

chromoxdor commented:
> I think it is better to have it more consistent, so please have a look to what is needed here.

I agree but it is a lot of work because there are a lot of colors that want to become variables....🙂
I look what i can do but i doubt that i´ll be finished with it this evening.

tonhuisman commented:
> Are there recent developments/changes of plugin commands?

It should be part of the regular development process, so mostly I'm waiting for this PR to be merged so I can start including it in my documentation process 😺

chromoxdor commented:
> It should be part of the regular development process, so mostly I'm waiting for this PR to be merged so I can start including it in my documentation process 😺

Ok... that reminds of doing a documentation here myself 😬

TD-er commented:
> > I think it is better to have it more consistent, so please have a look to what is needed here.
> 
> I agree but it is a lot of work because there are a lot of colors that want to become variables....🙂 I look what i can do but i doubt that i´ll be finished with it this evening.

Hmm if it will take "a lot of work" then maybe we should do it in a separate PR.
Please try to estimate the amount of work.
Point is right now it only will work on those who copy the files to their file system, so nobody is testing it.
Or I should make a separate CDN url for the code mirror stuff right now, but given the build size issues, I don't think that will make things easier right now.

chromoxdor commented:
i could offer this:
<img width="389" alt="Bildschirmfoto 2022-09-22 um 10 55 21" src="https://user-images.githubusercontent.com/33860956/191704261-a7e399b6-5bf3-4017-8ff3-12c80955695e.png">

instead of this:
<img width="389" alt="Bildschirmfoto 2022-09-22 um 10 53 37" src="https://user-images.githubusercontent.com/33860956/191704321-af2c92f2-7312-49f8-b9b9-d68c5baff70d.png">

> Hmm if it will take "a lot of work" then maybe we should do it in a separate PR.

Maybe we should. Right now it works fine in "auto" mode. 
Problem with me is, that i´ll be not available from tomorrow on until monday.  So i only have today....
Let´s say if i don´t make it until the end of the day i´ll make an extra PR


Edit: the picture above shows basically the dark mode with a different background which means a lot less colors to change to variables (most of the colors are the command specific ones) which means for me less writing and testing 😁

Edit2: To be clear, i am still totally fine with changing it completely...

TD-er commented:
> Let´s say if i don´t make it until the end of the day i´ll make an extra PR

Hmm that does sound a bit "definitive" ;)

Let's see how far you will get.
I am cleaning up the code a bit more, so I will also be pushing to this PR.

While cleaning up, I found a security bug.
Apparently it is possible to add `.src` to a filename in the URL and this will then be stripped off before trying to match it with a file on the file system.

Thus `/security.dat.src` is perfectly fine as a GET url and you will be able to download it without being logged in.
Not sure if anyone uses the `.src` trick, but it should at least check for credentials.
I guess it might even be better to remove it completely.

chromoxdor commented:
ok

chromoxdor commented:
> Let's see how far you will get.

ok

TD-er committed:
[CDN] Cleanup the CDN code

chromoxdor commented:
95% finished
I have to leave but i am confident, that i can finish it later that day...

chromoxdor committed:
Css variables update + add P023

chromoxdor commented:
100%... now i am late for my appointment....😳 (priorities 🤷🏻‍♂️)

TD-er committed:
[CDN] Browser cache  embedded Favicon and CSS

Serve embedded CSS and Favicon as separate file.
This way the file can be served with cache flags set and quickly re-served by sending a 304 (not modified) when it isn't.

TD-er committed:
Merge remote-tracking branch 'chromoxdor/esp_codemirror' into esp_codemirror

TD-er committed:
[Build] Fix build on ESP32

chromoxdor commented:
Safari seems to not play along....
![Bildschirmfoto 2022-09-22 um 16 38 38](https://user-images.githubusercontent.com/33860956/191778374-b7ab267e-51ed-4331-8965-71fdced7f5b9.png)

chromoxdor committed:
firefox was left out for "normal" rules, removed unnecessary folder

TD-er commented:
Can you show what the sent headers are when reloading the page?

N.B. the rules page is rendered differently on Firefox.
The text field is way too small.

chromoxdor commented:
> N.B. the rules page is rendered differently on Firefox.

what a timing... 🙂

TD-er commented:
> > N.B. the rules page is rendered differently on Firefox.
> 
> what a timing... 🙂

Ah I now see your commit :)

TD-er commented:
These code mirror files, should they be used on any other page other than the rules and maybe a custom page?

chromoxdor commented:
> Can you show what the sent headers are when reloading the page?
<img width="489" alt="Bildschirmfoto 2022-09-22 um 17 04 35" src="https://user-images.githubusercontent.com/33860956/191783471-6d82676e-371b-4b80-9647-daad3712ca8b.png">






> These code mirror files, should they be used on any other page other than the rules and maybe a custom page?

I guess not....... maybe in the future for the logs? but i have seen, that @Grovkillen has a very nice logviewer in his UI.

TD-er commented:
Safari does not send a `If-None-Match` request header?
![image](https://user-images.githubusercontent.com/3751318/191785571-6dcfeb94-40e8-4f64-ab57-c093eb1b7745.png)
Chrome and Firefox do send those when reloading a static file.

TD-er commented:
Apparently Safari does act differently on the caching.
Will try now what is suggested here: https://stackoverflow.com/a/36737536/8708166

TD-er committed:
[CDN] Only load codeMirror files on Rules/Custom page + add Expires

Needed for Safari

TD-er commented:
@chromoxdor 
Can you test with Safari?

chromoxdor commented:
Still the same:
![Bildschirmfoto 2022-09-22 um 17 45 06](https://user-images.githubusercontent.com/33860956/191792764-90e8d6fe-893f-4d8a-a5b6-8f84eb21538b.png)


On other pages the codemirror css is still loaded:
![Bildschirmfoto 2022-09-22 um 17 45 19](https://user-images.githubusercontent.com/33860956/191792923-8387c605-800b-4513-a9e8-8660c6c171fa.png)

TD-er commented:
I'm still working on the code mirror CSS...
But it is quite frustrating to see that Safari keeps loading the CSS.
Problem is that I may not be able to generate a proper Expire field as it needs to know the time. Thus if NTP is not working I may not be able to set a plausible timestamp.
I could perhaps base it on the build time when serving embedded code.

chromoxdor commented:
And if you now removing the the caching related stuff we are exactly at my proposal...🙂


> I had just a check in mind..like: if the page we visit is the rules page and we turned colorcode on, then put the files in the header.
> And since 99% of the user will most likely not store the files locally the lag in calling the rulespage once is negligible.

TD-er committed:
[CDN] Add Last-Modified header when serving built-in static files

To make Safari use caching

TD-er commented:
Please do check.
When serving built-in (embedded) CSS files, there should now be a Last-Modified header.
You may need to perform a full clean build to make it work though.... at least I hope as I don't see the header myself ....

chromoxdor commented:
i am not sure where i should see this.. this is what i get:
![Bildschirmfoto 2022-09-22 um 18 56 32](https://user-images.githubusercontent.com/33860956/191807469-b9590be6-cdc3-48c3-af44-b21d122b36b3.png)

chromoxdor commented:
By the way... putting the scrips after `initCM` and also using "defer" doesn´t work.. (defer always puts the scrips after parsing the html)
Scripts must be loaded before `initCM`.

chromoxdor commented:
Jep, just tested... defer must be removed and the scripts should be loaded before the textfield and before initCM(). (preferably in the head)
If the scripts load after the textfield than you´ll see the "normal" rules input for a fraction of a second before codemirror gets loaded.

TD-er commented:
Well that's easy to do as it is just a matter of changing a bool in the code where these are being called.

chromoxdor commented:
> Well that's easy

For you.. I actually wanted to be faster than you. But i am stuck in the code..i found the bool but couldn´t quite figure out where to change it...

TD-er commented:
> > Well that's easy
> 
> For you.. I actually wanted to be faster than you. But i am stuck in the code..i found the bool but couldn´t quite figure out where to change it...

In the `serve_JS` function:
```c++
        #if defined(WEBSERVER_INCLUDE_JS)
        if (!useCDN) {
          html_add_script_arg(id, true);
```
This last `true` should be replaced with `false` to switch all internally served JS without `defer`

That's the quickest test fix...

chromoxdor commented:
For testing i simply commented it out 😁
![Bildschirmfoto 2022-09-22 um 21 17 49](https://user-images.githubusercontent.com/33860956/191832684-52fb8555-c244-4f0e-a839-b34d3c36ff50.png)

But all other JS files can have the defer.. so it should be only the codemirror related JS without

chromoxdor committed:
reposition of codemirror js files + minor css additions

TD-er committed:
[CodeMirror] Link JS in the HEAD & don't use defer on those.

TD-er committed:
Merge remote-tracking branch 'chromoxdor/esp_codemirror' into esp_codemirror

TD-er commented:
I made the code mirror files be linked (without defer) in the head of the pages that need it.
It also now includes the `Last-Modified` header, so you should be able to use it in Safari too (given that that's the reason caching wasn't working)

chromoxdor commented:
Safari is somehow strange... first of all the 5 seconds waiting for loading a page. (this issue exists quite a while) and secondly caching still doesn't work 🤷🏻‍♂️
![Bildschirmfoto 2022-09-22 um 22 14 27](https://user-images.githubusercontent.com/33860956/191842862-beec0348-b2c5-4911-a3f3-15809c0af64d.png)

tonhuisman commented:
This PR can be linked to issue (Solves) #4086 so that will be closed once this is merged 👍

TD-er commented:
Load times of the rules page here (with 2k of rules included)
![image](https://user-images.githubusercontent.com/3751318/191846619-27767e52-05b8-4cd5-b8d5-9b5c6ad9a018.png)

So I guess Safari is a lost cause?

chromoxdor commented:
> So I guess Safari is a lost cause?

I think so 😁

TD-er commented:
> > So I guess Safari is a lost cause?
> 
> I think so 😁

Yep, as is stated by [Apple itself](https://support.apple.com/en-us/HT204416)

> Update Safari on Windows PC
> Apple no longer offers Safari updates for Windows. Safari 5.1.7 for Windows was the last version made for Windows, and it is now outdated.

chromoxdor commented:
Unfortunately i have an Iphone..so i can not run away from safari.
And there, every time i get an update, the browser renders my css a bit differently... its really annoying

TD-er commented:
Just to be sure, did you test with "embedded" CSS files and not served from a file on the filesystem?
It only sends the `Last-Modified` header for embedded files.

chromoxdor commented:
😬
<img width="454" alt="Bildschirmfoto 2022-09-22 um 23 40 43" src="https://user-images.githubusercontent.com/33860956/191856453-3d3fe361-87c1-4d62-8759-e7214ddbe392.png">

TD-er commented:
And for a refresh?
Does it then return a 304?

chromoxdor commented:
no..200 is all i get

TD-er commented:
Well at least Firefox and Chrome are doing it right.
Load times are now massively better than before.
![image](https://user-images.githubusercontent.com/3751318/191858481-9732614f-d682-46c7-a5cc-f9056328ad3e.png)

TD-er commented:
Anyway, can we merge the codemirror files?
Getting it to work with Apple's browser is something that shouldn't be depending on the availability of the JS/CSS files on the CDN servers.

chromoxdor commented:
> Anyway, can we merge the codemirror files?

i think we can...
